src/main/java/macrobase/analysis/MacrobaseAnalyzer.java

```diff
@@ -0,0 +1,47 @@
+package macrobase.analysis;
+
+import macrobase.analysis.outlier.OutlierClassifier;
+import macrobase.analysis.result.AnalysisResult;
+import macrobase.analysis.result.OutlierClassificationResult;
+import macrobase.analysis.summary.Summarizer;
+import macrobase.analysis.summary.Summary;
+import macrobase.conf.MacroBaseConf;
+import macrobase.datamodel.Datum;
+import macrobase.ingest.DataIngester;
+import macrobase.transform.FeatureTransformation;
+
+import java.util.List;
+
+public abstract class MacrobaseAnalyzer {
+
+    private DataIngester ingester;
+    private FeatureTransformation transformation;
+    private OutlierClassifier outlierClassifier;
+    private Summarizer summarizer;
+
+    public MacrobaseAnalyzer(MacroBaseConf conf) {};
+
+    public AnalysisResult analyze() {
+        for (Datum nextDatum = ingester.getNextDatum(); ; nextDatum = ingester.getNextDatum()) {
+            // While there are data points coming.
+            if (nextDatum == null) {
+               break;
+            }
+            // Apply transform(s) to the datum that we got from our engine.
+            Datum transformedDatum = transformation.transform(nextDatum);
+	        // Add transformed datum to the outlier classifier (could also be a
+	        // contextual classifier). This will only update the
+	        // classificationResults once per window, for performance optimization.
+            OutlierClassificationResult classificationResult = outlierClassifier.addDatum(transformedDatum);
+            // Summarizer will summarize outliers.
+	        // This will actually perform any computation only if the
+	        // classificationResult has been changed.
+            List<Summary> summaries = summarizer.summarize(classificationResult);
```

Comment:
What about breaking this down into two:

`void observe(Datum d)`

`Summary summarize()`